pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	// Don't modify the informer's copy.
 	tr := original.DeepCopy()
 
+	// build creation and status update adapted from prow: https://github.com/kubernetes/test-infra/blob/8faa040989cf8552683dcdd4400505ca62a3c408/prow/cmd/build/controller.go#L207
+	// create a build?
+	var wantBuild bool
+
+	switch {
+	case tr.DeletionTimestamp == nil:
+		wantBuild = true
+	}
+
+	// Get related task for taskrun
+	t, err := c.PipelineClientSet.PipelineV1alpha1().Tasks(tr.Namespace).Get(
+		tr.Spec.TaskRef.Name, metav1.GetOptions{})
+
+	var haveBuild bool
+
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case errors.IsNotFound(err):
+		// Do not have a build
+	case err != nil:
+		return fmt.Errorf("get build: %v", err)
+	case b.DeletionTimestamp == nil:
+		haveBuild = true
+	}
+
+	// Should we create or delete this build?
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			fmt.Printf("Deleted %s", key)
+			return nil
+		}
+		switch or := metav1.GetControllerOf(b); {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
+		}
+		fmt.Printf("Delete builds/%s", key)
+		if err = c.deleteBuild(namespace, name); err != nil {
+			return fmt.Errorf("delete build: %v", err)
+		}
+		return nil
+	case finalState(tr.Status.State):
+		fmt.Printf("Finished %s", key)
+		return nil
+	case wantBuild && t.Spec.BuildSpec == nil:
+		return fmt.Errorf("nil BuildSpec")
+	case wantBuild && !haveBuild:
+		if b, err = makeBuild(*t); err != nil {
+			return fmt.Errorf("make build: %v", err)
+		}
+		fmt.Printf("Create builds/%s", key)
+		if b, err = c.createBuild(namespace, b); err != nil {
+			return fmt.Errorf("create build: %v", err)
+		}
+	}
+
+	// Ensure taskrun status is correct
+	haveState := tr.Status.State
+	haveMsg := tr.Status.Description
+	wantState, wantMsg := taskRunStatus(b.Status)
+	if haveState != wantState || haveMsg != wantMsg {
+		ntr := tr.DeepCopy()
+		if ntr.Status.StartTime.IsZero() {
+			ntr.Status.StartTime = c.now()
+		}
+		if ntr.Status.CompletionTime.IsZero() && finalState(wantState) {
+			now := c.now()
+			ntr.Status.CompletionTime = &now
+		}
+		ntr.Status.State = wantState
+		ntr.Status.Description = wantMsg
+	}
+
 	// Reconcile this copy of the task run and then write back any status
 	// updates regardless of whether the reconciliation errored out.
 	err = c.reconcile(ctx, tr)
```

Comment:
I was expecting the actual logic to check/create the build to be in this private func, see example here: https://github.com/knative/serving/blob/76500a6917ce4aec522651677dd9bac677c6034f/pkg/reconciler/v1alpha1/revision/revision.go#L314:22
otherwise consider if we need this func?

Reply:
the way that @pivotal-nader-ziada 's example code explicitly expresses phases, then delegates work to them seems pretty slick!

Reply:
yes, I realize now that I put this in the wrong place (Reconcile vs reconcile).  I will move everything to the correct `reconcile` function, thanks for catching this!